### PR TITLE
[ao][sparsity] make sparsity and PTQ compose

### PR DESCRIPTION
**Pull request by HDCharles**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74848
* #74847
* #74846
* __->__ #74845

Summary: This PR adds support for quantization flow to detect
parametrized modules and match them using their original module types.
This mainly involved using the new type_before_parametrizations function rather than
type to check for module mathcing

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240274](https://our.internmc.facebook.com/intern/diff/D35240274)

### CODE REVIEW

File: torch/ao/quantization/utils.py [hunk 1]
```diff
@@ -360,3 +361,26 @@ def _parent_name(target):
         return '', r[0]
     else:
         return r[0], r[1]
+
+def nonparam_type(module):
```
Comment (vkuzo) [inline torch/ao/quantization/utils.py:365]: can we call it something like `type_before_parametrizations`?

File: torch/ao/quantization/utils.py [hunk 2]
```diff
@@ -360,3 +361,26 @@ def _parent_name(target):
...
+    """
+    Returns type(module) or the original
+    type if module is currently parametrized
+    """
+    if is_parametrized(module):
+        return module.__class__.__bases__[0]
+    else:
+        return type(module)
+
+def is_leaf(module):
```
Comment (vkuzo) [inline torch/ao/quantization/utils.py:375]: `is_leaf` is already used in the quantization codebase for the FX use case (a leaf is a module not traced through), could we pick a different name for this?

Comment (HDCharles) [reply torch/ao/quantization/utils.py:375]: yeah, I can do something like parametrized_check_for_leaf or something like that?

File: torch/ao/quantization/utils.py (hunk 1 above)
Comment (HDCharles) [reply torch/ao/quantization/utils.py:365]: sure

File: torch/ao/quantization/utils.py (hunk 2 above)
Comment (HDCharles) [reply torch/ao/quantization/utils.py:375]: i changed it to is_leaf_or_only_parametrized

Comment (vkuzo) [reply torch/ao/quantization/utils.py:375]: how about something like `has_no_children_ignoring_parametrizations`?

Comment (HDCharles) [reply torch/ao/quantization/utils.py:375]: ok

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240274).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240274).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240274).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240274).

File: test/ao/sparsity/test_composability.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        tq.prepare(mod, inplace=True)
+        sparsifier.prepare(mod, config=sparse_config)
+        self._check_parametrizations_and_observers(mod)
+        self._squash_mask_calibrate_and_convert(
+            mod, sparsifier, torch.randn(1, 4, 4, 4)
+        )
+        self.assertTrue(isinstance(mod[5], torch.nn.quantized.Linear))
+        self.assertEqual(mod(torch.randn(1, 4, 4, 4)).shape, torch.Size([1, 4, 4, 4]))
+
+    def test_s_prep_before_q_prep(self):
```
Comment (jerryzh168) [inline test/ao/sparsity/test_composability.py:80]: nit: maybe we can add a bit more context to this test in docstring about what this is testing

**Review by jerryzh168 — APPROVED** [review]

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240274).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240274).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240274).

**Pull request closed without merging**